File: src/main/java/com/dh/ondot/schedule/infra/dto/OdsayErrorResponse.java
package com.dh.ondot.schedule.infra.dto;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public record OdsayErrorResponse(
        @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
        List<Error> error
) {
    public record Error(
            String code,
            String message
    ) {
        @JsonCreator
        public Error(@JsonProperty("code") String code,
                     @JsonProperty("message") String message,
                     @JsonProperty("msg") String msg) {
            this(code, message != null ? message : msg);
        }
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/infra/dto/OdsayErrorResponse.kt
package com.dh.ondot.schedule.infra.dto

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty

data class OdsayErrorResponse(
    @param:JsonFormat(with = [JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY])
    val error: List<Error>,
) {
    data class Error(
        val code: String,
        val message: String,
    ) {
        companion object {
            @JvmStatic
            @JsonCreator
            fun create(
                @JsonProperty("code") code: String,
                @JsonProperty("message") message: String?,
                @JsonProperty("msg") msg: String?,
            ): Error = Error(code, message ?: msg ?: "")
        }
    }
}
